src/clj/muuntaja/core.clj

```diff
@@ -82,8 +82,9 @@
                 ;; optional guard on empty imput
                 (if (and allow-empty-input-on-decode? decode?)
                   (fn [^InputStream is charset]
-                    (if (pos? (.available is)) ($ is charset)))
-                  $))
+                    (if (and (not (nil? is)) (pos? (.available is))) ($ is charset)))
+                  (fn [^InputStream is charset]
+                    (if-not (nil? is) ($ is charset)))))
```

Comment:
There have been many reported problems from users when a 3rd party middleware has consumed the (sadly mutable) InputStream, leaving it empty, but not `nil`. Here's the explanation from compojure-api wiki: https://github.com/metosin/compojure-api/wiki/Serving-static-resources#caveats

Also, at least ring-jetty pushes empty inputstream instead of `nil` if there is no request body.

```clj
(def app
  (fn [request]
    (println (:body request))
    (println (.available (:body request)))))
; #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x6426abf4 HttpInputOverHTTP@6426abf4]
; 0
```

I couldn't figure out if the `.available` is always fast for all InputStreams. So I made it optional. If it's fast for all streams, it could be checked for all requests.

But as by Transit Spec - empty input is not allowed. I would like to support this, kind of "stric-mode". Proposal:

- `nil` input is turned into a empty Inputstream in `as-input-stream`, (object allocation, fast)
- remove `nil?` guard from `create-coder`
- default `:allow-empty-input-on-decode?` to true (fast for those adapter who return `nil`, most likely for others too
- setting it to `false` will make Muuntaja work "correct".

also, if there is a better name for the options, please tell.